relayer/relayer.go

```diff
@@ -50,41 +53,75 @@ func (r *Relayer) Start(ctx context.Context) error {
 	if err != nil {
 		r.logger.Fatalf("Error subscribing to new rollup block headers: %s", err.Error())
 	}
-
-	ticker := time.NewTicker(1500 * time.Millisecond)
-
-	var blocks []*ethtypes.Block
+	defer sub.Unsubscribe()
+
+	blocksToSubmit := make(chan []*ethtypes.Block)
+
+	ticker := time.NewTicker(SUBMIT_BLOCK_INTERVAL)
+	defer ticker.Stop()
+
+	go func() {
+		var blocks []*ethtypes.Block
+		for {
+			select {
+			case err := <-sub.Err():
+				r.logger.Errorf("error on rollup block subscription: %s", err.Error())
+				return
+			case header := <-headers:
+				r.logger.Info("Received rollup block header", "number", header.Number.Uint64())
+				blockWithNoTransactions := ethtypes.NewBlockWithHeader(header)
+				blocks = append(blocks, blockWithNoTransactions)
+			case <-ticker.C:
+				if len(blocks) > 0 {
+					blocksToSubmit <- blocks
+					blocks = nil
+					ticker.Stop()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
 
 	for {
 		select {
-		case err := <-sub.Err():
-			r.logger.Errorf("error on rollup block subscription: %s", err.Error())
-		case header := <-headers:
-			blockWithNoTransactions := ethtypes.NewBlockWithHeader(header)
-			blocks = append(blocks, blockWithNoTransactions)
-		case <-ticker.C:
-			if len(blocks) == 0 {
-				continue
+		case blocks := <-blocksToSubmit:
+			blockNumbers := make([]uint64, len(blocks))
+			for i, block := range blocks {
+				blockNumbers[i] = block.Number().Uint64()
 			}
+			r.logger.Info("Submitting blocks to NEAR", "numbers", blockNumbers)
 
 			encodedBlocks, err := rlp.EncodeToBytes(blocks)
 			if err != nil {
-				r.logger.Errorf("error RLP encoding block: %s", err.Error())
+				r.logger.Error("Error RLP encoding block", "err", err.Error())
 				continue
 			}
 
-			blocks = nil
+			out, err := r.nearClient.ForceSubmit(encodedBlocks)
+			if err != nil {
+				r.logger.Error("Error submitting block to NEAR", "err", err)
+
+				if strings.Contains(err.Error(), "InvalidNonce") {
+					r.logger.Info("Invalid nonce, resubmitting")
+					time.Sleep(SUBMIT_BLOCK_RETRY_TIMEOUT)
```

Comment:
Notice, If this part triggered it will delay ticker.Reset(SUBMIT_BLOCK_INTERVAL) on corresponding interval. Feels like there shall be a submission queue where element popped when it is successfully submitted. Otherwise we increate SUBMIT_BLOCK_INTERVAL on  SUBMIT_BLOCK_RETRY_TIMEOUT

Reply:
The problem here is if there was an invalid nonce, it means the SUBMIT_BLOCK_INTERVAL was not enough of a wait, so it makes sense to wait some more for the nonce update.
In the end, I don't think it matters having a big queue - if we wait too long to submit, the operators will ignore it either way, which is what I think would happen at this point with 2 intervals + 1 retry. So IMO not worth the extra hassle.